bottlerocket/agents/src/bin/vsphere-k8s-cluster-resource-agent/vsphere_k8s_cluster_provider.rs

```diff
@@ -0,0 +1,763 @@
+    pub vcenter_secret_name: Option<SecretName>,
+}
+
+impl Configuration for ProductionMemo {}
+
+/// Once we have fulfilled the `Create` request, we return information about the cluster we've created
+#[derive(Clone, Debug, Serialize, Deserialize, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct CreatedVSphereK8sCluster {
+    /// The name of the cluster.
+    pub cluster_name: String,
+
+    /// The cluster server endpoint.
+    pub endpoint: String,
+
+    /// Base64 encoded Kubeconfig for the vSphere K8s cluster
+    pub encoded_kubeconfig: String,
+}
+
+impl Configuration for CreatedVSphereK8sCluster {}
+
+pub struct VSphereK8sClusterCreator {}
+
+#[async_trait::async_trait]
+impl Create for VSphereK8sClusterCreator {
+    type Config = VSphereK8sClusterConfig;
+    type Info = ProductionMemo;
+    type Resource = CreatedVSphereK8sCluster;
+
+    async fn create<I>(
+        &self,
+        spec: Spec<Self::Config>,
+        client: &I,
+    ) -> ProviderResult<Self::Resource>
+    where
+        I: InfoClient,
+    {
+        let mut memo: ProductionMemo = client
+            .get_info()
+            .await
+            .context(Resources::Unknown, "Unable to get info from info client")?;
+        // Keep track of the state of resources
+        let mut resources = Resources::Clear;
+
+        // Get vSphere credentials to authenticate to vCenter via govmomi
+        let secret_name = spec
+            .secrets
+            .get(VSPHERE_CREDENTIALS_SECRET_NAME)
+            .context(resources, "Unable to fetch vSphere credentials")?;
+        vsphere_credentials(client, secret_name, &resources).await?;
+        memo.vcenter_secret_name = Some(secret_name.clone());
+
+        // Set current directory to somewhere other than '/' so eksctl-anywhere won't try to mount
+        // it in a container.
+        fs::create_dir_all(WORKING_DIR).context(
+            resources,
+            format!("Failed to create working directory '{}'", WORKING_DIR),
+        )?;
+        env::set_current_dir(Path::new(WORKING_DIR)).context(
+            resources,
+            format!("Failed to change current directory to {}", WORKING_DIR),
+        )?;
+
+        // Check whether cluster creation is necessary
+        let (do_create, message) = is_cluster_creation_required(
+            &spec.configuration.cluster,
+            spec.configuration.creation_policy.unwrap_or_default(),
+        )
+        .await?;
+        memo.current_status = message;
+        info!("{}", memo.current_status);
+        client
+            .send_info(memo.clone())
+            .await
+            .context(resources, "Error sending cluster creation message")?;
+
+        let mgmt_kubeconfig_path = format!("{}/mgmt.kubeconfig", WORKING_DIR);
+        let encoded_kubeconfig = if do_create {
+            // Check CAPI management cluster is accessible and valid
+            debug!("Decoding and writing out kubeconfig for the CAPI management cluster");
```

Comment:
This agent requires a CAPI management cluster to already exist?

Reply:
Yes, like you mentioned in the review message, I wanted to avoid bootstrapping with kind since the docker daemon is running with `vfs` as the underlying storage driver and it's incredibly inefficient since each container layer is its own directory. And setting up a kind bootstrap cluster requires a lot of containers to be running. I can try again to see if this is acceptable though.

Setting up overlayfs requires installing docker on the testsys node host and mounting in `/var/lib/docker` into the dind container which I find unappetizing as well.

By having an external management cluster, I don't have to worry about all of that. In addition, the management cluster becomes a source of truth for the state of all provisioned clusters and I can retrieve a kubeconfig for any provisioned cluster as long as I have access to the mgmt cluster. There are cons with this approach though. 

- I believe this makes integrating with `cargo make test` harder since the resource agent is no longer self-contained and requires an external kubeconfig (@ecpullen probably has opinions on this). 
- Someone needs to maintain the management cluster and remember to upgrade the CAPI, CAPV CRDs when EKS-A bumps versions.

Reply:
Makes sense. I think this is a great solution and we can consider "Don't require an external CAPI cluster" to be a separate issue/feature-request.

Reply:
The additional kubeconfig will either need to have a know location in the monorepo (like `testsys.kubeconfig`) or we will have to include it (as a string) in the `Test.toml` file. I prefer the former option, but not having to worry about it would be nice.

Reply:
I tried without a external management cluster and let EKS-A set up the bootstrap cluster in kind (kind running in docker that's running in docker btw lol) and it failed due to networking issues:
```
2022-10-25T20:35:28.490Z        V4      ----------------------------------
2022-10-25T20:35:28.490Z        V4      Task start      {"task_name": "bootstrap-cluster-init"}
2022-10-25T20:35:28.490Z        V0      Creating new bootstrap cluster
2022-10-25T20:35:28.491Z        V4      Creating kind cluster   {"name": "br-eksa-122-eks-a-cluster", "kubeconfig": "br-eksa-122/generated/br-eksa-122.kind.kubeconfig"}
2022-10-25T20:35:28.491Z        V6      Executing command       {"cmd": "/usr/bin/docker exec -i eksa_1666730106969714044 kind create cluster --name br-eksa-122-eks-a-cluster --kubeconfig br-eksa-122/generated/br-eksa-122.kind.kubeconfig --image public.ecr.aws/eks-anywhere/kubernetes-sigs/kind/node:v1.22.15-eks-d-1-22-11-eks-a-19 --config br-eksa-122/generated/kind_tmp.yaml"}
ERRO[2022-10-25T20:35:53.428755287Z] Could not add route to IPv6 network fc00:f853:ccd:e793::1/64 via device br-3443489d9de0: network is down 
time="2022-10-25T20:35:56.630430522Z" level=info msg="loading plugin \"io.containerd.event.v1.publisher\"..." runtime=io.containerd.runc.v2 type=io.containerd.event.v1
time="2022-10-25T20:35:56.630486720Z" level=info msg="loading plugin \"io.containerd.internal.v1.shutdown\"..." runtime=io.containerd.runc.v2 type=io.containerd.internal.v1
time="2022-10-25T20:35:56.630493149Z" level=info msg="loading plugin \"io.containerd.ttrpc.v1.task\"..." runtime=io.containerd.runc.v2 type=io.containerd.ttrpc.v1
time="2022-10-25T20:35:56.630582130Z" level=info msg="starting signal loop" namespace=moby path=/run/docker/containerd/daemon/io.containerd.runtime.v2.task/moby/99c8a19b51e1e85305f365229030c3b6b7e58f7423f120b88f02d0162ee1de7e pid=2752 runtime=io.containerd.runc.v2
```

So unless we want to invest time investigating how to get kind to work under 3 layers of container abstraction, I think we should stick with external mgmt cluster for now.

Reply:
:turtle: :turtle: :turtle: :turtle: :turtle: